File: src/python/algorithms/array/trappingRainWater/trapping_rain_water.py
# Approach 1: Two Pointers — O(n) time, O(1) space
def trapping_rain_water1(height):
    if not height:
        return 0
    left, right = 0, len(height) - 1
    left_max, right_max = height[0], height[right]
    res = 0
    while left < right:
        if left_max <= right_max:
            left += 1
            left_max = max(left_max, height[left])
            res += left_max - height[left]
        else:
            right -= 1
            right_max = max(right_max, height[right])
            res += right_max - height[right]
    return res


# Approach 2: Prefix/Suffix Arrays — O(n) time, O(n) space
def trapping_rain_water2(height):
    n = len(height)
    if n == 0:
        return 0
    left_max = [0] * n
    right_max = [0] * n
    left_max[0] = height[0]
    for i in range(1, n):
        left_max[i] = max(left_max[i - 1], height[i])
    right_max[n - 1] = height[n - 1]
    for j in range(n - 2, -1, -1):
        right_max[j] = max(right_max[j + 1], height[j])
    water = 0
    for i in range(n):
        water += min(left_max[i], right_max[i]) - height[i]
    return water


# Approach 3: Brute Force — O(n²) time, O(1) space
def trapping_rain_water3(height):
    n = len(height)
    if n == 0:
        return 0
    water = 0
    for i in range(n):
        left_max = max(height[:i + 1])
        right_max = max(height[i:])
        water += min(left_max, right_max) - height[i]
    return water


if __name__ == "__main__":
    t1 = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]
    t2 = [4, 2, 0, 3, 2, 5]
    t3 = [1, 1]
    t4 = []
    t5 = [3, 0, 0, 2, 0, 4]

    # Approach 1: Two Pointers
    r1 = trapping_rain_water1(t1)
    print(f"Test 1 (two-ptr): {'[PASS]' if r1 == 6 else '[FAIL]'} result={r1} expected=6")
    r2 = trapping_rain_water1(t2)
    print(f"Test 2 (two-ptr): {'[PASS]' if r2 == 9 else '[FAIL]'} result={r2} expected=9")
    r3 = trapping_rain_water1(t3)
    print(f"Test 3 (two-ptr): {'[PASS]' if r3 == 0 else '[FAIL]'} result={r3} expected=0")
    r4 = trapping_rain_water1(t4)
    print(f"Test 4 (two-ptr): {'[PASS]' if r4 == 0 else '[FAIL]'} result={r4} expected=0")
    r5 = trapping_rain_water1(t5)
    print(f"Test 5 (two-ptr): {'[PASS]' if r5 == 10 else '[FAIL]'} result={r5} expected=10")

    # Approach 2: Prefix/Suffix Arrays
    r6 = trapping_rain_water2(t1)
    print(f"Test 1 (prefix): {'[PASS]' if r6 == 6 else '[FAIL]'} result={r6} expected=6")
    r7 = trapping_rain_water2(t2)
    print(f"Test 2 (prefix): {'[PASS]' if r7 == 9 else '[FAIL]'} result={r7} expected=9")
    r8 = trapping_rain_water2(t3)
    print(f"Test 3 (prefix): {'[PASS]' if r8 == 0 else '[FAIL]'} result={r8} expected=0")
    r9 = trapping_rain_water2(t4)
    print(f"Test 4 (prefix): {'[PASS]' if r9 == 0 else '[FAIL]'} result={r9} expected=0")
    r10 = trapping_rain_water2(t5)
    print(f"Test 5 (prefix): {'[PASS]' if r10 == 10 else '[FAIL]'} result={r10} expected=10")

    # Approach 3: Brute Force
    r11 = trapping_rain_water3(t1)
    print(f"Test 1 (brute): {'[PASS]' if r11 == 6 else '[FAIL]'} result={r11} expected=6")
    r12 = trapping_rain_water3(t2)
    print(f"Test 2 (brute): {'[PASS]' if r12 == 9 else '[FAIL]'} result={r12} expected=9")
    r13 = trapping_rain_water3(t3)
    print(f"Test 3 (brute): {'[PASS]' if r13 == 0 else '[FAIL]'} result={r13} expected=0")
    r14 = trapping_rain_water3(t4)
    print(f"Test 4 (brute): {'[PASS]' if r14 == 0 else '[FAIL]'} result={r14} expected=0")
    r15 = trapping_rain_water3(t5)
    print(f"Test 5 (brute): {'[PASS]' if r15 == 10 else '[FAIL]'} result={r15} expected=10")

File: src/javascript/algorithms/array/trappingRainWater/trappingRainWater.js
// Approach 1: Two Pointers — O(n) time, O(1) space
function trappingRainWater1(height) {
    if (height.length === 0) return 0;
    let left = 0, right = height.length - 1;
    let leftMax = height[0], rightMax = height[right];
    let res = 0;
    while (left < right) {
        if (leftMax <= rightMax) {
            left++;
            leftMax = Math.max(leftMax, height[left]);
            res += leftMax - height[left];
        } else {
            right--;
            rightMax = Math.max(rightMax, height[right]);
            res += rightMax - height[right];
        }
    }
    return res;
}

// Approach 2: Prefix/Suffix Arrays — O(n) time, O(n) space
function trappingRainWater2(height) {
    const n = height.length;
    if (n === 0) return 0;
    const leftMax = new Array(n);
    const rightMax = new Array(n);
    leftMax[0] = height[0];
    for (let i = 1; i < n; i++) leftMax[i] = Math.max(leftMax[i - 1], height[i]);
    rightMax[n - 1] = height[n - 1];
    for (let j = n - 2; j >= 0; j--) rightMax[j] = Math.max(rightMax[j + 1], height[j]);
    let water = 0;
    for (let i = 0; i < n; i++) water += Math.min(leftMax[i], rightMax[i]) - height[i];
    return water;
}

// Approach 3: Brute Force — O(n²) time, O(1) space
function trappingRainWater3(height) {
    const n = height.length;
    if (n === 0) return 0;
    let water = 0;
    for (let i = 0; i < n; i++) {
        let leftMax = 0;
        for (let j = 0; j <= i; j++) leftMax = Math.max(leftMax, height[j]);
        let rightMax = 0;
        for (let j = i; j < n; j++) rightMax = Math.max(rightMax, height[j]);
        water += Math.min(leftMax, rightMax) - height[i];
    }
    return water;
}

const t1 = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1];
const t2 = [4, 2, 0, 3, 2, 5];
const t3 = [1, 1];
const t4 = [];
const t5 = [3, 0, 0, 2, 0, 4];

// Approach 1: Two Pointers
const r1 = trappingRainWater1(t1);
console.log(`Test 1 (two-ptr): ${r1 === 6 ? '[PASS]' : '[FAIL]'} result=${r1} expected=6`);
const r2 = trappingRainWater1(t2);
console.log(`Test 2 (two-ptr): ${r2 === 9 ? '[PASS]' : '[FAIL]'} result=${r2} expected=9`);
const r3 = trappingRainWater1(t3);
console.log(`Test 3 (two-ptr): ${r3 === 0 ? '[PASS]' : '[FAIL]'} result=${r3} expected=0`);
const r4 = trappingRainWater1(t4);
console.log(`Test 4 (two-ptr): ${r4 === 0 ? '[PASS]' : '[FAIL]'} result=${r4} expected=0`);
const r5 = trappingRainWater1(t5);
console.log(`Test 5 (two-ptr): ${r5 === 10 ? '[PASS]' : '[FAIL]'} result=${r5} expected=10`);

// Approach 2: Prefix/Suffix Arrays
const r6 = trappingRainWater2(t1);
console.log(`Test 1 (prefix): ${r6 === 6 ? '[PASS]' : '[FAIL]'} result=${r6} expected=6`);
const r7 = trappingRainWater2(t2);
console.log(`Test 2 (prefix): ${r7 === 9 ? '[PASS]' : '[FAIL]'} result=${r7} expected=9`);
const r8 = trappingRainWater2(t3);
console.log(`Test 3 (prefix): ${r8 === 0 ? '[PASS]' : '[FAIL]'} result=${r8} expected=0`);
const r9 = trappingRainWater2(t4);
console.log(`Test 4 (prefix): ${r9 === 0 ? '[PASS]' : '[FAIL]'} result=${r9} expected=0`);
const r10 = trappingRainWater2(t5);
console.log(`Test 5 (prefix): ${r10 === 10 ? '[PASS]' : '[FAIL]'} result=${r10} expected=10`);

// Approach 3: Brute Force
const r11 = trappingRainWater3(t1);
console.log(`Test 1 (brute): ${r11 === 6 ? '[PASS]' : '[FAIL]'} result=${r11} expected=6`);
const r12 = trappingRainWater3(t2);
console.log(`Test 2 (brute): ${r12 === 9 ? '[PASS]' : '[FAIL]'} result=${r12} expected=9`);
const r13 = trappingRainWater3(t3);
console.log(`Test 3 (brute): ${r13 === 0 ? '[PASS]' : '[FAIL]'} result=${r13} expected=0`);
const r14 = trappingRainWater3(t4);
console.log(`Test 4 (brute): ${r14 === 0 ? '[PASS]' : '[FAIL]'} result=${r14} expected=0`);
const r15 = trappingRainWater3(t5);
console.log(`Test 5 (brute): ${r15 === 10 ? '[PASS]' : '[FAIL]'} result=${r15} expected=10`);
